### DISCUSSION TITLE
Connection timed out even with ~ 10 Registers?

### DISCUSSION THREAD

**Discussion by gutschein** [aldas/modbus-tcp-client#85] [Q&A] [answered]
Hi,

my target are ~ 30 Registers, but I started with just a few. Even with more than ~8 Registers I run into Connection timed out.
I tried to increase wait time, but maybe I added the 2nd parameter in a wrong pattern? 

```
example: $response = (new NonBlockingClient(['readTimeoutSec' => 0.5]))->sendRequests($fc3);
I Tried: $response = (new NonBlockingClient(['readTimeoutSec' => 5.0],['ConnectTimeoutSec' => 5.0]))->sendRequests($fc3);
```
Is this correct? What can I do else?

**Comment by aldas** [comment, marked as answer]
Options array should be created like that

```php
$client = new NonBlockingClient([
    'readTimeoutSec' => 1.5, // timeout when waiting response from server
    'connectTimeoutSec' => 1.5,  // timeout when establishing connection to the server
]);
$responseContainer = ($client)->sendRequests($fc3);
```

How many requests are sent? `$x = count($fc3);` are all sent to same IP, same unitID?

I recommend trying to send same requests  (same start addres + quantity + ip + unitID) with https://www.modbusdriver.com/modpoll.html and see how it behaves. It could be that your devices behaves.

**Comment by gutschein** [comment]
thank you, now the paramter works! :)

Good hint with the count: It seems to me, that my device answers really quick up to 4 requests (even w/o increasing the timeout value). with a 5th I get a connect timeout (even with 10sec timeout)
I tried to order the registers by addresses and this reduces the count of requests with the same count of registers. But sometimes I reach the 5th request anyway.

How are the requests are build? Can I control the count of addresse per request?

**Reply by aldas** [reply]
Request are built by grouping addresses into groups by  ip+unitID+address range (consecutive 124 registers)

https://github.com/aldas/modbus-tcp-client/blob/1fb85de630c45154c015f779c0785987d54c70e1/src/Composer/AddressSplitter.php#L24

If you can - group addresses you read so they there are no caps between. 30 addresses should not be that much registers (1 register = 2 bytes).  FC3 request can returns 124 registers max by 1 one request . 

My experience for modbus devices, in my case Wago PLC's,  is that about ~10 requests seems to be limit and then device seems to timeout and behave unreliably.  so depending on the device - experience and limits could be different.

**Reply by aldas** [reply]
if your request interval requirements allows (time between is large enough) - you could send some requests serially and not parallel.

**Comment by gutschein** [comment]
Ok, good informantion. I just request 1 IP+ID: I've a gap beween almost any address. In any case it's not much data at all, but spreaded...
I tried to sort the addresse and reached now 24 of my needed registers in 4 requests. But with the next address (with gap) leading to a 5th request I get the timeout.

Just for an impression for you: e.g. 14 Addresse (each 4Bytes) lead to 5 requests... which seems to be a magic limit of my device.
```
30595 
30597
30775
30845
30849
30877
30879
30881
30883
30917
31009
31057
34109
40187
```
* Is this splitting a modbus-definition or could we customize the way of splitting? Or could it be solved in the library in a different way?
* If I try split it at any gap I come to more than 5 requests. Small gaps are allowed? Or means 124 Registers ~496 Bytes starting at the first address and it's splitted after the last address fitting in this intervall? Next Intervall starts with next request address?

Even with sortingen I may have to split it up into chunks and request them seperatly. Trying serial. My target was just one call per IP+ID with many requests but maybe it's not possible that way.

**Reply by aldas** [reply]
Modbus request can returns only ~250bytes. this is ~124 registers. So when this list of addresses is split into request  you have groups/requests like that:

Request	|	start address	|	start + max quantity	|	diff to next
--- | --- | --- | -- 
1	|	30595	|	30719	|	2
1	|	30597	|		|	178
2	|	30775	|	30899	|	70
2	|	30845	|		|	4
2	|	30849	|		|	28
2	|	30877	|		|	2
2	|	30879	|		|	2
2	|	30881	|		|	2
2	|	30883	|		|	34
3	|	30917	|	31041	|	92
4	|	31009	|	31133	|	48
4	|	31057	|		|	3052
5	|	34109	|	34233	|	6078
6	|	40187	|	40311	|	


as you see there are big caps between addresses and therefore new request is needed

**Comment by gutschein** [comment]
I solved it by now with chunks - even with timeout-limit = 0.5... as long as I stay below request_max = 5:
```
$sortby  = array_column($registers, 'address');
array_multisort($sortby, SORT_ASC, $registers);

$fc3 = ReadRegistersBuilder::newReadHoldingRegisters()
    ->allFromArray($registers)
    ->build();

$request_max = 4;
$request_wait = 100;
$fc3_chunks = array_chunk($fc3, $request_max);
print_r('Request Chunks: '.count($fc3_chunks));echo("\n");
$i = 1;
foreach($fc3_chunks as $fc3_part) {
        print_r('Requests in Chunk '.$i.' : '.count($fc3_part));echo("\n");
        $response = (new NonBlockingClient(['readTimeoutSec' => 0.5, 'connectTimeoutSec' => 0.5]))->sendRequests($fc3_part);
        print_r($response->getData());
        $i++;
        usleep($request_wait*1000);
}
```
```
Registers: 34
Requests at all: 8
Request Chunks: 2
```

I would appreciate to get answers to my message before anyway. Maybe there are better options for device specific request restrictions... :)

**Comment by gutschein** [comment]
ok. Understood. Thank you. 

Maybe it makes sense to implement the request sorting (to keep request count as small as possible) and the chunks into the library?
I'm totally fine without it (means: I'm really happy how it works now 😄  ) as it works since some hours well with sorting + chunking by myself. 

But maybe that would be more handsome and an inprovement for "Register-Junkies"...

**Reply by aldas** [reply]
It is already included https://github.com/aldas/modbus-tcp-client/blob/1fb85de630c45154c015f779c0785987d54c70e1/src/Composer/AddressSplitter.php#L24 is responsible to splitting addresses to requests.

What you do is - after addresses are splitted to reasonable sized (start address + quantity) request  you chunk them so reasonable amount of requests are made in parallel  and excess serially chunks

**Reply by aldas** [reply]
My recommendation is - try to keep all addresses consecutive to each other. start with 30597 and make sure that next value you want to read is  30597+2 value size in registers.  For example that 40187 is so far off other addresses that splitter will create separate requests just for that address. if it would be nearer to other addresses you would have less requests.
